lib/ProgramCall.js

```diff
@@ -31,59 +33,108 @@ class ProgramCall {
+   * https://github.com/tc39/proposal-private-methods
+   * We should update to use private methods in the future.
+   * @param {string} name
+   * @param {string} type
+   * @param {string} value
+   * @param {object} options
+  */
+  __addDataNode__(name, type, value, options) {
+    const dataName = name ? ` name='${name}'` : '';
+    this.xml += `<data${dataName} type='${type}'`;
+    // append <data> options
+    if (typeof options === 'object') {
+      if (options.varying) { this.xml += ` varying='${options.varying.varying}'`; }
+      if (options.enddo) { this.xml += ` enddo='${options.enddo}'`; }
+      if (options.setlen) { this.xml += ` setlen='${options.setlen}'`; }
+      if (options.offset) { this.xml += ` offset='${options.offset}'`; }
+      if (options.hex) { this.xml += ` hex='${options.hex}'`; }
+      if (options.trim) { this.xml += ` trim='${options.trim}'`; }
+      if (options.next) { this.xml += ` next='${options.next}'`; }
     }
-    if (!inDs) { // In recursive mode, if it is an element in DS, then no <parm> or </parm> needed.
-      this.xml += '<parm';
-      if (opt && typeof opt === 'object') { // append <param> options
-        if (opt.io) { this.xml += ` io='${opt.io}'`; }
-        if (opt.by) { this.xml += ` by='${opt.by}'`; }
+    this.xml += `>${value}</data>`;
+  }
+
+  /**
+   * Internal function to create <ds> nodes
+   * There is an open propsal to add private methods to JS.
+   * https://github.com/tc39/proposal-private-methods
+   * We should update to use private methods in the future.
+   * @param {array} ds
+   * @param {object} options
+   */
+  __addDsNodes__(ds, options = {}) {
+    if (Array.isArray(ds[0])) { // check if we are dealing with a 2D array
+      for (let i = 0; i < ds.length; i += 1) {
+        if (!Array.isArray(ds[i][0])) {
+          // handles 2D array without nested DS
+          this.__addDataNode__(ds[i][0], ds[i][1], ds[i][2], ds[i][3]);
+          // eslint-disable-next-line no-continue
+          continue;
+        }
+        // handle nested ds
+        this.xml += '<ds';
+        // append <ds> options
+        if (options.dim) { this.xml += ` dim='${options.dim}'`; }
+        if (options.dou) { this.xml += ` dou='${options.dou}'`; }
+        if (options.len) { this.xml += ` len='${options.len}'`; }
+        if (options.data) { this.xml += ` data='${options.data}'`; }
+        this.xml += '>';
+
+        for (let j = 0; j < ds[i].length; j += 1) {
+          this.__addDsNodes__(ds[i][j]);
+        }
+        this.xml += '</ds>';
       }
-      this.xml += '>';
+    } else {
+      this.__addDataNode__(ds[0], ds[1], ds[2], ds[3]);
     }
+  }
 
-    if (Array.isArray(data)) { // If it is a struct parameter, recursively parse its children.
+  /**
+   * Internal function to handle ds and data within <parm> node
+   * @param {object} parameter
+   */
+  __addData__(parameter = {}) {
+    // adding a data structure with data nodes
+    if (Array.isArray(parameter.value)) {
```

Comment:
Why are we still handling arrays here? Let `__addDsNodes__` handle this. Also, I thought we were switching to object-only anyway.

Reply:
Oh, never mind, I see that we are just checking if it's a ds or not. That should be done by checking `parameter.type == 'ds'` instead.

Reply:
Ok will update to check for `parameter.type == 'ds'`